pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -456,3 +487,315 @@ func (plugin *kubenetNetworkPlugin) getNsenterPath() (string, error) {
 	}
 	return plugin.nsenterPath, nil
 }
+
+type closeable interface {
+	Close() error
+}
+
+type hostport struct {
+	port     int
+	protocol string
+}
+
+type targetPod struct {
+	podFullName string
+	podIP       string
+}
+
+func (hp *hostport) String() string {
+	return fmt.Sprintf("Port: %s, Protocol: %s", hp.port, hp.protocol)
+}
+
+//validatePodHostports validates hostport specified in pod if any
+func (plugin *kubenetNetworkPlugin) validatePodHostports(pod *api.Pod) error {
+	for _, container := range pod.Spec.Containers {
+		for _, port := range container.Ports {
+			hp := hostport{
+				port:     int(port.HostPort),
+				protocol: strings.ToLower(string(port.Protocol)),
+			}
+			if _, ok := plugin.hostPortMap[hp]; ok {
+				return fmt.Errorf("Hostport %v specified in pod %s is already been taken. ", hp, kubecontainer.GetPodFullName(pod))
+			}
+		}
+	}
+	return nil
+}
+
+func (plugin *kubenetNetworkPlugin) gatherAllHostports() (map[api.ContainerPort]targetPod, error) {
+	podHostportMap := make(map[api.ContainerPort]targetPod)
+	pods, err := plugin.host.GetRuntime().GetPods(false)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to retrieve pods from runtime: %v", err)
+	}
+	for _, p := range pods {
+		var podInfraContainerId kubecontainer.ContainerID
+		for _, c := range p.Containers {
+			if c.Name == dockertools.PodInfraContainerName {
+				podInfraContainerId = c.ID
+				break
+			}
+		}
+		cidr, ok := plugin.podCIDRs[podInfraContainerId]
+		if !ok {
+			// The POD has been delete. Ignore
+			continue
+		}
+		podIP, _, err := net.ParseCIDR(strings.Trim(cidr, "\n"))
+		if err != nil {
+			glog.Errorf("Failed to retrieve pod ip for %s-%s: %v", p.Namespace, p.Name, err)
+			continue
+		}
+		// Need the complete api.Pod object
+		pod, ok := plugin.host.GetPodByName(p.Namespace, p.Name)
+		if ok {
+			for _, container := range pod.Spec.Containers {
+				for _, port := range container.Ports {
+					if port.HostPort != 0 {
+						podHostportMap[port] = targetPod{podFullName: kubecontainer.GetPodFullName(pod), podIP: podIP.String()}
+					}
+				}
+			}
+		}
+	}
+	return podHostportMap, nil
+}
+
+// Join all words with spaces, terminate with newline and write to buf.
+func writeLine(buf *bytes.Buffer, words ...string) {
+	buf.WriteString(strings.Join(words, " ") + "\n")
+}
+
+func hostportChainName(cp api.ContainerPort, podFullName string) utiliptables.Chain {
+	hash := sha256.Sum256([]byte(string(cp.HostPort) + string(cp.Protocol) + podFullName))
+	encoded := base32.StdEncoding.EncodeToString(hash[:])
+	return utiliptables.Chain(kubenetHostportChainPrefix + encoded[:16])
+}
+
+// syncHostportsRules r
+func (plugin *kubenetNetworkPlugin) syncHostportsRules() {
+	start := time.Now()
+	defer func() {
+		glog.V(4).Infof("syncHostportsRules took %v", time.Since(start))
+	}()
+
+	containerPortMap, err := plugin.gatherAllHostports()
```

Comment:
Do you need to do this periodically, or can we save it?  Is it even right to be doing this in kubenet?  Not all plugins will be able to offer host ports, but let's ignore that  If we have a system with 2 different networks with different drivers, and both do support hostPorts, shouldn't we coordinate between them?  Maybe this checking should be at a higher level?  Or maybe we shouldn't do it at all - it's always possible that a system daemon has the requested port, in which case we have to handle failure anyway.